Add copy/cut/paste with header

boutinb opened the pull request:
Fixes https://github.com/jasp-stats/INTERNAL-jasp/issues/2375

If a copy is done from the column header, the column name is also copied in the clipboard, so that if the user paste it in Excel, then the whole column with its name is copied. 
If a paste is done in a cell in JASP, then the clipboard is copied. 
If a paste is done in a column header, then if the copy was done form a column header, then the whole column (with name, labels, description...) is copied. 
If a cut is done in a column header, the whole column is removed.

Moreover it ensures now that the name of a column is unique.

boutinb committed:
Add copy/paste with header

If a copy is done from the column header, the column name is also copied in the clipboard, so that if the user paste it in Excel, then the whole column with its name is copied.
If a paste is done in a cell in JASP, then the clipboard is copied.
If a paste is done in a column header, then if the copy was done form a column header, then the whole column (with name, labels, description...) is copied.
If a cut is done in a column header, the whole column is removed.

Moreover it ensures now that the name of a column is unique.

JorisGoosen commented:
> Moreover it ensures now that the name of a column is unique.

it already did that in paste with header

```
void DataSetPackage::unicifyColumnNames()
{
	for(int c=0; c<dataColumnCount(); c++)
	{
		std::string name = getColumnName(c);

		for(int c1=c+1; c1<dataColumnCount(); c1++)
			if(getColumnName(c1) == name)
			{
				std::string newName = name + "'";
				while(!isColumnNameFree(newName))
					newName = newName + "'";
				setColumnName(c1, newName);
			}
	}
}
```

but now we apparently have two functions for that haha

JorisGoosen commented on CommonData/column.cpp:
```diff
@@ -89,10 +89,11 @@ void Column::setName(const std::string &name)
 	if(_name == name)
 		return;
 
+	_name = getUniqueName(name);
```
This is quite dangerous btw...

You now can have the situation where you set the columnname to **X** from somewhere else and get confirmation that it changed but then it is actually called **X2** something

Could lead to problems at:
https://github.com/jasp-stats/jasp-desktop/blob/c7dbe8e2e3463364ce2fd082cddb96e2e61d7973/Desktop/data/datasetpackage.cpp#L1740-L1745

https://github.com/jasp-stats/jasp-desktop/blob/c7dbe8e2e3463364ce2fd082cddb96e2e61d7973/Desktop/data/importers/importer.cpp#L140-L141

https://github.com/jasp-stats/jasp-desktop/blob/c7dbe8e2e3463364ce2fd082cddb96e2e61d7973/Desktop/analysis/analysis.cpp#L480-L486

JorisGoosen replied:
Well, it turns out it mostly doesnt matter, because all of the places this gets called eventually already make sure that the columnnames are in fact unique.

JorisGoosen commented on Desktop/components/JASP/Widgets/DataTableView.qml:
```diff
@@ -257,15 +259,15 @@ FocusScope
 							if(controlPressed)
 							{
-								theView.cut(shiftPressed);
+								theView.cut(false);
```
Shouldnt you also do this for https://github.com/jasp-stats/jasp-desktop/blob/c7dbe8e2e3463364ce2fd082cddb96e2e61d7973/Desktop/components/JASP/Widgets/JASPDataView.qml#L86-L99 ?

And maybe also just remove the argument if we arent using it anymore

JorisGoosen commented on Desktop/data/columnmodel.cpp:
```diff
@@ -258,7 +283,8 @@ void ColumnModel::setChosenColumn(int chosenColumn)
 	DataSet * data = DataSetPackage::pkg()->dataSet();
 	
-	subNodeModel()->selectNode(data ? data->column(chosenColumn) : nullptr);
+	_virtual = data ? (chosenColumn >= data->columnCount()) : true;
```
```suggestion
	_virtual = !data || data->columnCount() <= chosenColumn;
```

JorisGoosen commented on CommonData/column.cpp:
```diff
@@ -89,10 +89,11 @@ void Column::setName(const std::string &name)
 
+	_name = getUniqueName(name);
+
 	if(_title.empty() || _title == _name)
```
You broke this check by overriding `_name`...

So before when you would change (lets say) `V1` to `Tsjakka` the title would change with it, but now it doesnt...

JorisGoosen commented:
So changes requested:
- columntitle should stay the same as the columnname unless the user specifically edited it
- JASPDataView should behave the same as the main dataviewer thing
- There shouldnt be multiple functions doing the same thing in JASP. So Either just use the existing "UnicifyColumnNames" (Why did you not use it @boutinb ? ) or remove it entirely if it isnt necessary anymore.

boutinb committed:
Make code a bit clearer

Implement comments from Joris: https://github.com/jasp-stats/jasp-desktop/pull/5249#issuecomment-1719685040

pasteSpreadSheet does not change the col names anymore.

Cut row should remove row if done from row header.

boutinb commented:
> So changes requested:
> 
> * columntitle should stay the same as the columnname unless the user specifically edited it
> * JASPDataView should behave the same as the main dataviewer thing
> * There shouldnt be multiple functions doing the same thing in JASP. So Either just use the existing "UnicifyColumnNames" (Why did you not use it @boutinb ? ) or remove it entirely if it isnt necessary anymore.

I think I have added most of your comments.

Also cut rows should removed rows when done from row header.

JorisGoosen commented:
Sorry, its almost good but:
<img width="626" alt="image" src="https://github.com/jasp-stats/jasp-desktop/assets/29062713/d2af8bd3-330f-494b-8f73-d2ee69fd949d">

I changed the name for `contExpon` to `contExpon!!!` (title changed with it), but when copying it it you get `contExpon!!! 2` as name but it keeps the title as is... Sorry, label or long name I mean!

boutinb committed:
Title should be stay the same as name, if it was so before copying.
